src/cell/unsafe_cell.rs

```diff
@@ -214,6 +227,13 @@ impl<T> From<T> for UnsafeCell<T> {
     }
 }
 
+impl<T: ?Sized> Drop for UnsafeCell<T> {
+    fn drop(&mut self) {
+        // Observe that we have mutable access.
+        self.get_mut();
```

Comment:
Should `std::mem::needs_drop()` be used to solve this issue:
> You restrict the type's use relative to std's UnsafeCell. When Drop is implemented, drop check applies, making some uses that worked before no longer work.


```suggestion
        // Observe that we have mutable access if the inner type needs to be dropped.
        if std::mem::needs_drop::<T>() {
            self.get_mut();
        }
```

Reply:
I don't agree with this.

1. Even if the inner type doesn't need to be dropped and thus its drop implementation does nothing, dropping the cell while a reference to the inner type exists is still wrong.

2. The documentation for needs_drop says that "this is purely an optimization hint". But we're changing observable behavior here. Using it would lead to safety assertions changing whether they get triggered based on the return value of needs_drop, which can change.

Reply:
> Even if the inner type doesn't need to be dropped and thus its drop implementation does nothing, dropping the cell while a reference to the inner type exists is still wrong.

Actually, this is isn't wrong to me. If an object is dropped, it means that all references pointing to it have reach the end of their lifetime, otherwise, the code wouldn't even compile.
In fact, after thinking a second time about this sentence, it means that #349 is not an issue after all, and doesn't need to be "fixed".

Reply:
> The documentation for needs_drop says that "this is purely an optimization hint". But we're changing observable behavior here.

This is an optimization hint, but the documentation still states "However if this function actually returns false, then you can be certain dropping T has no side effect". So `needs_drop` result is still guaranteed to have no false-negative, and can be relied on. 
Of course, as written in the documentation, it could lead to false-positives, a few false positive may be still better than a too much conservative behavior leading to more false-positives. And if the result of `needs_drop` is changing with compiler versions, it's still guaranteed to not introduce false-negative.

Reply:
Anyway, see my [comment](https://github.com/tokio-rs/loom/issues/349#issuecomment-2227472851) in the issue, I don't loom is doing things right here. Dropping a cell should not be counted a mutable access but as a cell invalidation, for further `ConstPtr`/`MutPtr` dereference.